# Basic Data Analysis Project

### PART 1, Basic Python

Read `long-beach.txt` file and convert all the contents similar to `long-beach-output.txt`, where only two records are processed 

In [55]:
#with open("../datasets/long-beach.txt","r") as file:
 #   print(file.read())

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

raw_text = None
with open("../datasets/long-beach.txt", "r") as f:
    raw_text = f.read().replace("\n","")

entries = raw_text.split("name")
with open("../datasets/long-beach-useroutput.txt", "w") as f:
    for e in entries:
        f.write(f"{e}name\n")



In [56]:
# Your code here


## Part 2, Exploratory Data Analysis using Pandas

I provide you a data file named `data.csv`. You need to perform exploratory data analysis. I have also created some checkpoints in csv files like `checkpoint1.csv`, so that you can load the corresponding `.csv` file to achieve the processing done till that point.

In [57]:
# import necessary libraries here 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Read the `data.csv` file and display the contents as shown below

<img src="../images/Fig1.PNG" />

In [58]:
# Your code here
df = pd.read_csv("../datasets/data.csv")
df.head()

Jobclass                Job  age Education-level    marital-status   race  \
0    Private  Machine-op-inspct   25            11th     Never-married  Black   
1    Private    Farming-fishing   38         HS-grad  Married-A-spouse  White   
2  Local-gov    Protective-serv   28      Assoc-acdm  Married-A-spouse  White   
3    Private  Machine-op-inspct   44    Some-college  Married-A-spouse  Black   
4          ?                  ?   18    Some-college     Never-married  White   

   gender  hours-per-week native-country income  
0    Male              40  United-States  <=50K  
1    Male              50  United-States  <=50K  
2    Male              40  United-States   >50K  
3    Male              40  United-States   >50K  
4  Female              30  United-States  <=50K

How many records are there in `data.csv`?

In [59]:
# Your code here 
df.shape[0]

48842

Total 48842 rows and 10 columns in the dataset

In `data.csv`, there are some missing values, which are denoted by `?`. List down all the columns with number of missing values as shown below

<img src="../images/Fig2.PNG" />

In [60]:
# Your code here 
missing_values = (df == "?").sum()
missing_values

Jobclass           2799
Job                2809
age                   0
Education-level       0
marital-status        0
race                  0
gender                0
hours-per-week        0
native-country      857
income                0
dtype: int64

Interesting column `Job` and column `Jobclass` contains many missing values. How many records have both these columns missing?

In [61]:
# Your code here 
print(f"Job is missing {missing_values['Job']}")
print(f"Jobclass is missing {missing_values['Jobclass']}")

Job is missing 2809
Jobclass is missing 2799


Replace all `?` occurrences with `NaN`

In [62]:
# Your code here 
df[df=="?"] = np.nan
df

Jobclass                Job  age Education-level    marital-status  \
0           Private  Machine-op-inspct   25            11th     Never-married   
1           Private    Farming-fishing   38         HS-grad  Married-A-spouse   
2         Local-gov    Protective-serv   28      Assoc-acdm  Married-A-spouse   
3           Private  Machine-op-inspct   44    Some-college  Married-A-spouse   
4               NaN                NaN   18    Some-college     Never-married   
...             ...                ...  ...             ...               ...   
48837       Private       Tech-support   27      Assoc-acdm  Married-A-spouse   
48838       Private  Machine-op-inspct   40         HS-grad  Married-A-spouse   
48839       Private       Adm-clerical   58         HS-grad           Widowed   
48840       Private       Adm-clerical   22         HS-grad     Never-married   
48841  Self-emp-inc    Exec-managerial   52         HS-grad  Married-A-spouse   

        race  gender  hours-per-week native-country income  
0      Black    Male              40  United-States  <=50K  
1      White    Male              50  United-States  <=50K  
2      White    Male              40  United-States   >50K  
3      Black    Male              40  United-States   >50K  
4      White  Female              30  United-States  <=50K  
...      ...     ...             ...            ...    ...  
48837  White  Female              38  United-States  <=50K  
48838  White    Male              40  United-States   >50K  
48839  White  Female              40  United-States  <=50K  
48840  White    Male              20  United-States  <=50K  
48841  White  Female              40  United-States   >50K  

[48842 rows x 10 columns]

### Checkpoint1

List unique enteries in columns `Job`, `Jobclass` and `native-country`

In [63]:
# Your code here 
print(f"for Job: {df.Job.unique()}\n")
print(f"for Jobclass: {df.Jobclass.unique()}\n")
print(f"for native-country: {df['native-country'].unique()}\n")

for Job: ['Machine-op-inspct' 'Farming-fishing' 'Protective-serv' nan
 'Other-service' 'Prof-specialty' 'Craft-repair' 'Adm-clerical'
 'Exec-managerial' 'Tech-support' 'Sales' 'Priv-house-serv'
 'Transport-moving' 'Handlers-cleaners' 'Armed-Forces']

for Jobclass: ['Private' 'Local-gov' nan 'Self-emp-not-inc' 'Federal-gov' 'State-gov'
 'Self-emp-inc' 'Without-pay' 'Never-worked']

for native-country: ['United-States' nan 'Peru' 'Guatemala' 'Mexico' 'Dominican-Republic'
 'Ireland' 'Germany' 'Philippines' 'Thailand' 'Haiti' 'El-Salvador'
 'Puerto-Rico' 'Vietnam' 'South' 'Columbia' 'Japan' 'India' 'Cambodia'
 'Poland' 'Laos' 'England' 'Cuba' 'Taiwan' 'Italy' 'Canada' 'Portugal'
 'China' 'Nicaragua' 'Honduras' 'Iran' 'Scotland' 'Jamaica' 'Ecuador'
 'Yugoslavia' 'Hungary' 'Hong' 'Greece' 'Trinadad&Tobago'
 'Outlying-US(Guam-USVI-etc)' 'France' 'Holand-Netherlands']



Replace `NaN` in columns `Job`, `Jobclass` and `native-country` with the most frequently occurred value.

Hint: Statistical function `Mode` can help in this

In [64]:
# df_tmp[pd.isna(df_tmp["Job"])]["Job"] = "hello"
df_tmp = df[["Job", "Jobclass", "native-country"]].copy()  # Create an explicit copy
df_tmp.loc[pd.isna(df_tmp["Job"]), "Job"] = df_tmp["Job"].mode().iloc[0]
df_tmp.loc[pd.isna(df_tmp["Jobclass"]), "Jobclass"] = df_tmp["Jobclass"].mode().iloc[0]
df_tmp.loc[pd.isna(df_tmp["native-country"]), "native-country"] = df_tmp["native-country"].mode().iloc[0]
df.loc[:, ["Job", "Jobclass", "native-country"]] = df_tmp

df

Jobclass                Job  age Education-level    marital-status  \
0           Private  Machine-op-inspct   25            11th     Never-married   
1           Private    Farming-fishing   38         HS-grad  Married-A-spouse   
2         Local-gov    Protective-serv   28      Assoc-acdm  Married-A-spouse   
3           Private  Machine-op-inspct   44    Some-college  Married-A-spouse   
4           Private     Prof-specialty   18    Some-college     Never-married   
...             ...                ...  ...             ...               ...   
48837       Private       Tech-support   27      Assoc-acdm  Married-A-spouse   
48838       Private  Machine-op-inspct   40         HS-grad  Married-A-spouse   
48839       Private       Adm-clerical   58         HS-grad           Widowed   
48840       Private       Adm-clerical   22         HS-grad     Never-married   
48841  Self-emp-inc    Exec-managerial   52         HS-grad  Married-A-spouse   

        race  gender  hours-per-week native-country income  
0      Black    Male              40  United-States  <=50K  
1      White    Male              50  United-States  <=50K  
2      White    Male              40  United-States   >50K  
3      Black    Male              40  United-States   >50K  
4      White  Female              30  United-States  <=50K  
...      ...     ...             ...            ...    ...  
48837  White  Female              38  United-States  <=50K  
48838  White    Male              40  United-States   >50K  
48839  White  Female              40  United-States  <=50K  
48840  White    Male              20  United-States  <=50K  
48841  White  Female              40  United-States   >50K  

[48842 rows x 10 columns]

In [65]:
# Second Method


### Checkpoint2

`income` column contains strings `<=50K` and `>50K`. Replace with `0` and `1` respectively.

In [74]:
# Your code here 
df.income = df.income.map(lambda x: 0 if "<=50" in str(x) else 1)

In [76]:
df

Jobclass                Job  age Education-level    marital-status  \
0           Private  Machine-op-inspct   25            11th     Never-married   
1           Private    Farming-fishing   38         HS-grad  Married-A-spouse   
2         Local-gov    Protective-serv   28      Assoc-acdm  Married-A-spouse   
3           Private  Machine-op-inspct   44    Some-college  Married-A-spouse   
4           Private     Prof-specialty   18    Some-college     Never-married   
...             ...                ...  ...             ...               ...   
48837       Private       Tech-support   27      Assoc-acdm  Married-A-spouse   
48838       Private  Machine-op-inspct   40         HS-grad  Married-A-spouse   
48839       Private       Adm-clerical   58         HS-grad           Widowed   
48840       Private       Adm-clerical   22         HS-grad     Never-married   
48841  Self-emp-inc    Exec-managerial   52         HS-grad  Married-A-spouse   

        race  gender  hours-per-week native-country  income  
0      Black    Male              40  United-States       0  
1      White    Male              50  United-States       0  
2      White    Male              40  United-States       1  
3      Black    Male              40  United-States       1  
4      White  Female              30  United-States       0  
...      ...     ...             ...            ...     ...  
48837  White  Female              38  United-States       0  
48838  White    Male              40  United-States       1  
48839  White  Female              40  United-States       0  
48840  White    Male              20  United-States       0  
48841  White  Female              40  United-States       1  

[48842 rows x 10 columns]

List down the data types of every column

In [77]:
#Your code here
df.dtypes

Jobclass           object
Job                object
age                 int64
Education-level    object
marital-status     object
race               object
gender             object
hours-per-week      int64
native-country     object
income              int64
dtype: object

Convert column `income` data type to `int32`

In [79]:
#Your code here 
df.income = df.income.astype("int32")

In [81]:
df.dtypes

Jobclass           object
Job                object
age                 int64
Education-level    object
marital-status     object
race               object
gender             object
hours-per-week      int64
native-country     object
income              int32
dtype: object

### Checkpoint3

Plot the heat map of the corelation between all the numeric columns in the dataframe as shown below

<img src="../images/Fig3.PNG" />

In [ ]:
# Your code here


The heatmap shows that there is a relation between the `income` and `age` column. Plot a box-plot as shows in the figure below

<img src="../images/Fig4.PNG" />

In [ ]:
#Your code here 


The `Education-level` and `income` are usually related. Plot a bar chart that shows the count of people who have a certain level in education and have income `<=50K` and `>50K` as shows below in figure

<img src="../images/Fig5.PNG" />

In [ ]:
#Your code here 


#### Bonus question 
Plot a similar plot as above where instead of `count`, it gives the `percentage` as shown in figure below

<img src="../images/Fig6.PNG" />

Similar plot for `Job` and `income` as shown in figure below

<img src="../images/Fig7.PNG" /> 

In [ ]:
#Your code here 


#### Bonus question 
Plot a similar plot as above where instead of `count`, it gives the `percentage` as shown in figure below

<img src="../images/Fig8.PNG" />

In [ ]:
#Your code here


There are 3 enteries of `marital-status` where the person is married (Married-A-spouse, Married-B-spouse and Married-C-spouse). Make a new column `Married`, that is `1` where a person is married otherwise '0' as shown in figure below

<img src="../images/Fig9.PNG" />

In [ ]:
#Your code here 


Plot a bar that counts the persons that have income `<=50K` and `>50K` and married or not as shown in figure below.

<img src="../images/Fig10.PNG" />

Write a one line, what this plot describes

In [ ]:
#Your code here 
